### ISSUE TITLE
Allow "raw" URL opening.

### ISSUE THREAD

**Issue by kevincox** [tridactyl/tridactyl#1002]
If I try `:open foobar` this searches for "foobar" using my default search engine.

I would like to have an option to instead navigate to `http://foobar`. This is similar to setting `keyword.enabled=false`.

I tried `set searchurls.browse http://%s` however this appends a `/` after the URL which breaks some things.

This could possibly be done with a `-r` option to the `*open` functions.

**Comment by kevincox** [comment]
In my case I need a large enough number that that approach isn't feasible. My current workaround is using <C-l>to focus the native URL bar.

**Comment by glacambre** [comment]
This issue still exists, but while rewriting some of the code related to URL-handling I noticed that you could specify the website's port in order to force your query to be treated as a url, e.g. `tabopen foobar:80` should work. Still not optimal, but probably better than creating thousands of searchurls :)

**Comment by glacambre** [comment]
Just realized that you can also do `:set searchurls. d http://%s` to have `open d foobar` work.

**Comment by vlmarek** [comment]
FWIF, I had similar (just reverse - I needed to open strings in search engine) issue way back with pentadactyl. I have solved it back then by

// https://github.com/5digits/dactyl/blob/master/common/content/dactyl.js#L1075

javascript <<EOF
dactyl.urlish = util.regexp(
  "^(<domain>+\\.<domain>+[^0-9](:\\d+)?(/.*)?|localhost)$",
  "ix",
  { domain: "[^\\u0000-\\u002c\\u002f\\u003a-\\u0040\\u005b-\\u0060\\u007b-\\u007f]" }
);
EOF

compared to original, my change forces the string to contain dot in order to be treated as a domain. This allowed me to have strings like "data/123/456" to be passed to my search engine

**Comment by bovine3dom** [comment]
@vlmarek you can do that easily in Tridactyl with `open search [URL]` which is bound to `s` by default.

**Comment by vlmarek** [comment]
Not really, but close. You say to pass everything to a search engine. What I did in pentadactyl (and maybe is not necessary in tridactl - don't know yet) is to change which string is treated as url and opened directly and which string is forwarded to search engine.

On the other hand, the search engine in my case is my cgi which does all the logic (like if I `:open 12345` it opens corresponding bug, but if I `:open 12345678` it opens something else as bugs do not have so high numbers). So it could be distinguishing which URLs are to be opened directly. The inconvenience is that it would go always through my CGI and would be a little bit slower.

My understanding is that the original requester of this issue wanted the same - to have a way to change how the string is interpreted - whether it's direct url or call to search engine.

**Comment by bovine3dom** [comment]
`:js window.location.href="http://foobar"` is the most obvious way of doing this

**Comment by Snaipe** [comment]
Oh God, that's not a great workaround -- I certainly wouldn't bother typing all that to open http://foobar.

For what it's worth, this issue was what led me to completely abandon :open and friends in favor of simply using the default unified URL+search field that firefox provides. At least it does something sensible (i.e., search foobar if "foobar" is typed, and open http://foobar if "foobar/" is typed) when used.

**Comment by bovine3dom** [comment]
```
:command rawopen js -d¾ window.location.href=JS_ARGS.join(" ")¾
:rawopen http://foobar
```

this accidentally made me find https://foobar.withgoogle.com/ and now I want to know what it is

... I don't see why we couldn't just interpret a single word ending with a / as a url so I'll reopen this

**Comment by alanpearce** [comment]
> I don't see why we couldn't just interpret a single word ending with a / as a url so I'll reopen this

I would like to go a bit further than this, namely that `:open foo/bar` would open `http://foo/bar`. My use case is that I use [golink](https://github.com/tailscale/golink#golink) which is available at `http://go` and redirects e.g. `http://go/ex` to `http://example.com`[^1].

I thought that this might be a bit more controversial and/or require some discussion, so I didn't include it in my PR: the intent of `:open http/2` is more likely to be a search than a domain name lookup.

The ideal scenario here for me would be to have `:open` follow the browser fixup rules, which are configurable under `browser.fixup`: if I have `browser.fixup.dns_first_for_single_words` enabled, or `browser.fixup.domainsuffixwhitelist.go`, then `go` or `go/ex` are both interpreted as URLs, but `:open` interprets the latter as a search.

Some ideas for possible implementations:
 - lookup the `domainsuffixwhitelist` preferences or mirror them in config
 - a configuration item that makes `:open` fall back to opening the address as a URL instead of searching (since users like me can use `:open {search,$searchengine,$searchurl}` to specify the intent to search over visiting a literal URL)

I'd be happy to try to work on the implementation given some direction on what approach to take

[^1]: My current workaround is `set searchurls.go http://go/%s`, which requires I use a space instead of a slash in `:open`, which is okay, but not great.

**Comment by bovine3dom** [comment]
Since I wrote the previous workaround, we added `jsurls` which allow you to do e.g.

```
set jsurls.r url => "http://" + url
open r go/ex
tabopen r go/ex
winopen r go/ex
```

etc.

is that good enough for you?

if not, making the URL wrangling user configurable is probably reasonable. We could run a function here https://github.com/tridactyl/tridactyl/blob/d30d6d9df40d5b4b9a10e5490bd55fea195a8660/src/lib/webext.ts#L294 that mapped string -> string | undefined, where undefined means "continue with the default function" and string means go to that href

**Comment by alanpearce** [comment]
> is that good enough for you?

I think it might be. I replaced my *open bindings with your suggestion, now I just have to get into the habit of using s/S to search

**Comment by bovine3dom** [comment]
We could move jsurl handling further up that function and make returning `undefined` continue on with the other checks, deleting the jsurl

then you could put whitelist filter in or just check for spaces, whatever you wanted.

But I suspect that might be a feature only for you...

**Comment by alanpearce** [comment]
Yeah, no need for that 👍
